gnat2goto/driver/ureal_to_binary.adb

```diff
@@ -0,0 +1,228 @@
+with Uintp; use Uintp;
+with Types; use Types;
+with Ada.Text_IO;
+
+package body Ureal_To_Binary is
+   package IO renames Ada.Text_IO;
+
+   procedure Get_Integer_And_Fractional_Parts (Number : Ureal;
+                                              Int_Part : out String;
+                                              Frac_Part : out String;
+                                              Int_End : out Integer;
+                                              Frac_End : out Integer)
+     with Pre => (Denominator (Number) > Uint_0);
+
+   procedure Get_Integer_And_Fractional_Parts (Number : Ureal;
+                                               Int_Part : out String;
+                                               Frac_Part : out String;
+                                               Int_End : out Integer;
+                                               Frac_End : out Integer) is
+      Num : Uint := Numerator (Number);
+      Den : constant Uint := Denominator (Number);
+
+      function Get_Next_Exponent return Int;
+      procedure Do_Integer_Part;
+      procedure Do_Fractional_Part;
+
+      function Get_Next_Exponent return Int is
+         E : Int := 0;
+         Eth_Power_Of_Two : Uint := Uint_1;
+      begin
+         while (Den * Eth_Power_Of_Two) < Num loop
+            E := E + 1;
+            Eth_Power_Of_Two := Eth_Power_Of_Two * 2;
+         end loop;
+         return E;
+      end Get_Next_Exponent;
+
+      procedure Do_Integer_Part is
+         E : Int := Get_Next_Exponent;
+         Mantissa : Uint := Uint_2**E;
+         Digit_Position : Integer := Int_Part'First;
+      begin
+         --  we start with a string filled with 0
+         Int_Part := (others => '0');
```

Comment:
This is array initialisation syntax, for anyone not familiar with it (I wasn't, Hannes explained)